src/JsonSchema/Validator.elm

```diff
@@ -41,283 +41,371 @@ type ErrorMessage
     | DecodeError String
 
 
-validateObject : ObjectSchema -> Dict String Value -> List Error
-validateObject schema values =
-    List.concatMap (validateProperty values) schema.properties
+{-| Either data extracted from validated json, or a list of errors received while validating
+-}
+type ValidatedValue
+    = Invalid (List Error)
+    | StringValue String
+    | ArrayValue (Array.Array ValidatedValue)
+    | ObjectValue (Dict String ValidatedValue)
+    | BoolValue Bool
+    | FloatValue Float
+    | IntValue Int
+    | LazyValue (List Error)
+    | Valid Value
+
+
+{-| Validate a JSON Value against a schema.
+
+If validation fails, a list of errors is returned, otherwise the list is empty.
+
+It does not yet validate the `format` keyword.
+
+-}
+validate : Schema -> Value -> List Error
+validate schema value =
+    toErrors <| getValidatedValue [] schema value
+
+
+{-| This function validates your JSON against a schema.
+
+If validation succeeds, you can then gain access to the data you need
+
+If validation fails, you can extract a list of errors
+
+-}
+getValidatedValue : Pointer -> Schema -> Value -> ValidatedValue
+getValidatedValue pointer schema v =
```

Comment:
💭 Do you think we'd win anything by using [elm-verify](http://package.elm-lang.org/packages/stoeffel/elm-verify/latest) for this?